File: solutions/javascript/GenerateParentheses/GenerateParentheses.js
var res_arr = [];
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    res_arr = [];
    gars(n, n, '');
    return res_arr;
};
//get all result string
var gars = function(ln, rn, res) {
    if (ln == 0 && rn == 0) {
        res_arr.push(res);
    } else {
        if (ln > 0) {
            gars(ln - 1, rn, res + '(');
        }
        if (rn > 0 && rn > ln) {
            gars(ln, rn - 1, res + ')');
        }
    }
};

File: solutions/java/GenerateParentheses/GenerateParentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        getResult(result, n, n, "");
        return result;
    }

    private void getResult(List<String> list, int left, int right, String res) {
        if (left == 0 && right == 0) {
            list.add(res);
        } else {
            if (left > 0) {
                getResult(list, left - 1, right, res + "(");
            }
            if (right > 0 && left < right) {
                getResult(list, left, right - 1, res + ")");
            }
        }
    }
}